proto/p4/p4runtime.proto

```diff
@@ -399,11 +400,27 @@ message RegisterEntry {
   P4Data data = 3;
 }
 
+// Used to configure the digest extern only, not to stream digests or acks
+message DigestEntry {
+  uint32 digest_id = 1;
+  // a DigestList message is streamed when the following conditions are met:
+  //   - there is at least one digest ready
+  //   - the oldest digest in the list has been waiting for at least timeout_ns
+  //     nanoseconds or we have gathered max_list_size digests already
+  message Config {
+    int64 max_timeout_ns = 1;  // max timeout for outstanding digest data
+    int32 max_list_size = 2;  // max size for a digest list
+    int64 ack_timeout_ns = 3;  // timeout for DigestListAck message
```

Comment:
My original intention for the ack message was not to limit the number of messages "in-flight" to 1.

Rather the ack mechanism is to make sure that only a single digest notification is generated for the same data values. Let's take the example of MAC learning. Maybe I have generated a `DigestList` message for `{MAC1, MAC2}`. I don't want to keep generating messages for `MAC1` (for example) until the client has taken action, even if I keep receiving packets with source MAC set to `MAC1`. In the MAC learning example, the action taken by the client would probably be to update the smac & dmac tables, before sending the ack for the message.
This ack mechanism can greatly help reduce the amount of data sent IMO.

If you don't think this is a good / practical idea, we can probably discuss in-person (after the meeting tomorrow?).

Do you think there is an advantage in limiting the number of digests in-flight to a single one?

Reply:
Your message helps clarify the intent.  Thanks.

So the intent is that the server keeps a 'cache' containing the set of all digest message contents that have been sent, but not acknowledged, up to ack_timeout_ns in the past?  And the server deletes entries when they are at least ack_timeout_ns old, or the client sends a DigestListAck message for them?  And such a DigestListAck message effectively acknowledges _all_ digest messages that were grouped together into the same DigestList message with that (digest_id, list_id) pair?  If so, that all makes sense.  The ack_timeout_ns can be configured to 0 by a client if it wants to know about all digest messages, even if they are repeats.  It can configure it to 1 second if it wants to only hear repeated digest messages at least 1 second apart.  I've heard the name "deduplication FIFO" used to refer to this when implemented in an ASIC, although FIFO could be considered a misnomer there.

So this means that when a server receives a digest message from the local data plane, it must look up the message contents in this cache, e.g. an exact match hash table, and discard it if it is in that cache?  If so, also makes sense.

Having some description of this behavior, probably in the P4Runtime API spec document, would be handy for future reference.

A question for which I do not have a good answer at the moment, and maybe it is not a big deal at all, is what should be done with this cache if/when one controller takes over for another.  Emptying it out might be reasonable.

Another question that makes me think of is: If there are any cases where more than one client wants to get digest messages from the same data plane, is P4Runtime the way to do that, or should P4Runtime simplify its life by recommending that in such situations, the the server will only send them to one client,and that client is responsible for sharing it with any others who are interested?  Making a server responsible for sending to multiple clients directly sounds like a mess in how it would interact with this ack mechanism.

Reply:
Yes that's exactly what I was trying to express. My only concern is that it would probably be hard for an ASIC to implement this "deduplication FIFO" for the general case where more than one digest can be in-flight (i.e. not acknowledged) yet.

I agree that we need a detailed description of this in the spec. I'll add it to the draft as soon as we agree on the interface.

Regarding your other questions:
- we could empty the cache when mastership changes, although that is not strictly necessary (we can recover quickly once the cache expires)
- I think we still have to solve the matter of sending notifications to multiple controllers (all controllers for the relevant role: master + slaves). This is common to digest and packet-in. Even if we start sending digests to slaves, I believe only the master would be responsible for sending the ack.